File: 300_Longest_Increasing_Subsequence.java
class Solution {

    public int lengthOfLIS(int[] nums) {

        /* Sanity check */
        if(nums.length == 0)
            return 0;

        /* dp[i] array stores the smallest tail of increasing array with
         *      length of i + 1. It can be easily proved that this array
         *      is monotonically increasing.
         **/
        int[] dp = new int[nums.length];
        int bound = 0;
        dp[0] = nums[0];

        for(int i = 1; i < nums.length; ++i) {

            if(nums[i] > dp[bound])         /* A longer subsequence is found */
                dp[++bound] = nums[i];

            else {                          /* Binary search */
                int lo = 0, hi = bound;
                while (lo < hi) {

                    int mid = (lo + hi) / 2;
                    
                    if(nums[i] > dp[mid])
                        lo = mid + 1;
                    else
                        hi = mid;
                }
                
                dp[lo] = nums[i];
            }

        }

        return bound + 1;

    }

}

File: 300_Longest_Increasing_Subsequence.cpp
class Solution {
 public:
  int lengthOfLIS(vector<int>& nums) {
    int max = 1;
    if (nums.empty()) return 0;
    vector<int> length(nums.size(), 1);
    length[0] = 1;
    for (int i = 1; i < nums.size(); ++i) {
      for (int j = 0; j < i; ++j) {
        if (nums[i] > nums[j] && length[i] < length[j] + 1) {
          length[i] = length[j] + 1;
          if (length[i] > max) max = length[i];
        }
      }
    }
    return max;
  }
};

// 2023-12-27
class Solution2 {
public:
    int max(int a, int b) {
        return a > b ? a : b;
    }

    int lengthOfLIS(vector<int>& nums) {
        vector<int> dp(nums.size(), 1);

        for (int i = 1; i < nums.size(); i++) {
            for (int j = 0; j < i; j++) {
                if (nums[i] > nums[j]) {
                    dp[i] = max(dp[i], dp[j] + 1);
                }
            }
        }
        return *max_element(dp.begin(), dp.end());
    }
};